File: services/push_service/app/Services/CircuitBreaker.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Redis;
use App\Traits\JsonLogging;

/**
 * CircuitBreaker
 *
 * Manages the state of external service calls to prevent cascading failures.
 * Tracks failure count and transitions between CLOSED, OPEN, and HALF-OPEN states.
 * Logs all state changes using structured JSON logs.
 *
 * States:
 * - CLOSED: Normal operation
 * - OPEN: Calls are blocked
 * - HALF-OPEN: Limited calls allowed to test recovery
 *
 * Configuration is loaded from config/services.php
 */
class CircuitBreaker
{
    use JsonLogging;

    private string $serviceName;
    private int $failureThreshold;
    private int $timeout;
    private int $retryTimeout;

    /**
     * Constructor
     *
     * Initializes circuit breaker configuration.
     *
     * @param string $serviceName Name of the external service
     */
    public function __construct(string $serviceName)
    {
        $this->serviceName = $serviceName;
        $config = config('services.circuit_breaker');
        $this->failureThreshold = $config['failure_threshold'];
        $this->timeout = $config['timeout'];
        $this->retryTimeout = $config['retry_timeout'];
    }

    /**
     * Check if the circuit is open
     *
     * @return bool True if the circuit is open, false otherwise
     */
    public function isOpen(): bool
    {
        $state = Redis::get($this->getStateKey());

        if ($state === 'open') {
            $openedAt = Redis::get($this->getOpenedAtKey());

            if ($openedAt && (time() - $openedAt) > $this->retryTimeout) {
                $this->halfOpen();
                return false;
            }

            return true;
        }

        return false;
    }

    /**
     * Record a successful call
     *
     * Resets failure count and closes the circuit if necessary
     */
    public function recordSuccess(): void
    {
        Redis::del($this->getFailureCountKey());
        $this->close();
    }

    /**
     * Record a failed call
     *
     * Increments failure count and opens circuit if threshold is reached
     */
    public function recordFailure(): void
    {
        $failures = Redis::incr($this->getFailureCountKey());
        Redis::expire($this->getFailureCountKey(), $this->timeout);

        if ($failures >= $this->failureThreshold) {
            $this->open();
        }
    }

    /**
     * Open the circuit breaker
     *
     * Sets state to OPEN and logs the event
     */
    private function open(): void
    {
        Redis::setex($this->getStateKey(), $this->timeout, 'open');
        Redis::set($this->getOpenedAtKey(), time());

        $this->logJson(
            'error',
            $this->serviceName,
            'circuit_opened',
            "Circuit breaker OPENED for service {$this->serviceName}"
        );
    }

    /**
     * Close the circuit breaker
     *
     * Sets state to CLOSED and logs the event
     */
    private function close(): void
    {
        Redis::set($this->getStateKey(), 'closed');
        Redis::del($this->getOpenedAtKey());

        $this->logJson(
            'info',
            $this->serviceName,
            'circuit_closed',
            "Circuit breaker CLOSED for service {$this->serviceName}"
        );
    }

    /**
     * Set the circuit breaker to HALF-OPEN state
     *
     * Allows limited calls to test service recovery
     */
    private function halfOpen(): void
    {
        Redis::set($this->getStateKey(), 'half-open');

        $this->logJson(
            'warning',
            $this->serviceName,
            'circuit_half_open',
            "Circuit breaker HALF-OPEN for service {$this->serviceName}"
        );
    }

    /**
     * Get the Redis key for the circuit breaker state
     *
     * @return string
     */
    private function getStateKey(): string
    {
        return "circuit_breaker:{$this->serviceName}:state";
    }

    /**
     * Get the Redis key for the failure count
     *
     * @return string
     */
    private function getFailureCountKey(): string
    {
        return "circuit_breaker:{$this->serviceName}:failures";
    }

    /**
     * Get the Redis key for the timestamp when circuit was opened
     *
     * @return string
     */
    private function getOpenedAtKey(): string
    {
        return "circuit_breaker:{$this->serviceName}:opened_at";
    }
}

File: services/api_gateway/app/services/circuit_breaker.py
# ============================================
# api-gateway/app/services/circuit_breaker.py
# ============================================
import time
import logging
from enum import Enum
from app.config.settings import settings

logger = logging.getLogger(__name__)


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


class CircuitBreaker:
    def __init__(self):
        self.failure_count = 0
        self.failure_threshold = settings.CIRCUIT_BREAKER_FAILURE_THRESHOLD
        self.timeout = settings.CIRCUIT_BREAKER_TIMEOUT
        self.last_failure_time = None
        self.state = CircuitState.CLOSED

    async def call(self, func):
        """
        Execute function with circuit breaker protection
        """
        if self.state == CircuitState.OPEN:
            if self._should_attempt_reset():
                self.state = CircuitState.HALF_OPEN
                logger.info("Circuit breaker entering HALF_OPEN state")
            else:
                raise Exception("Circuit breaker is OPEN")

        try:
            result = await func()
            self._on_success()
            return result
        except Exception as e:
            self._on_failure()
            raise e

    def _on_success(self):
        """Handle successful call"""
        self.failure_count = 0
        if self.state == CircuitState.HALF_OPEN:
            self.state = CircuitState.CLOSED
            logger.info("Circuit breaker CLOSED")

    def _on_failure(self):
        """Handle failed call"""
        self.failure_count += 1
        self.last_failure_time = time.time()

        if self.failure_count >= self.failure_threshold:
            self.state = CircuitState.OPEN
            logger.warning("Circuit breaker OPENED")

    def _should_attempt_reset(self) -> bool:
        """Check if enough time has passed to try again"""
        return (
            self.last_failure_time
            and time.time() - self.last_failure_time >= self.timeout
        )
